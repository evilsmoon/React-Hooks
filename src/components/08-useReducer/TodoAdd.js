import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({ handelAddTodo }) => {

    const [{ description }, handelInputChange, reset] = useForm({
        description: ""
    })

    const handelSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }

        const newToDo = {
            id: new Date().getTime(),
            desc: description,
            done: false

        }
        handelAddTodo(newToDo);
        reset();

    }
    return (

        <>
            <h4>add ToDo</h4>
            <hr />
            <form onSubmit={handelSubmit}>
                <input
                    type='text'
                    name='description'
                    placeholder='add..'
                    autoComplete='off'
                    className='form-control'
                    value={description}
                    onChange={handelInputChange}
                />
                <button
                    type='submit'
                    className='btn btn-primary mt-1 btn-block'
                >
                    Add
               </button>
            </form>
        </>
    )
}
