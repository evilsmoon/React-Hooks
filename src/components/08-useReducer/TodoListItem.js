import React from 'react'
import './styles.css'

export const TodoListItem = ({ todo, index, handelToggle ,handelDelete}) => {
    // console.log(todo, i)
    return (
        <li
            key={todo.id}
            className='list-group-item'
        >
            <p
                className={`${todo.done && 'complete'} `}
                onClick={() => {
                    handelToggle(todo.id)

                }}
            >
                {index + 1}. {todo.desc}
            </p>
            <button
                className='btn btn-danger'
                onClick={() => {
                    handelDelete(todo.id)
                }}
            >
                delete
            </button>
        </li>
    )
}
