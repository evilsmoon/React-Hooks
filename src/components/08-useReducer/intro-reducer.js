
const initialState =[{
    id:1,
    todo:'buy weed',
    done:false
}]

const todoReducer = (state = initialState,action)=>{
   
   if(action?.type === 'add'){
       
       return [...state,action.payload];
   }
   
}
let todos = todoReducer();

// todos.push() no usar push en los arreglos por que muta

const newTodo = {
    id:2,
    todo:'smook weed',
    done:false
}

const addTodoAction = {
    type:'add',
    payload: newTodo
}
todos = todoReducer( todos,addTodoAction );
console.log(todos)

