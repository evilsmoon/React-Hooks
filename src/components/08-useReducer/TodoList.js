import React from 'react';
import { TodoListItem } from './TodoListItem';
export const TodoList = ({ todos ,handelDelete,handelToggle}) => {

    console.log(todos)
    return (
        <ul className='list-group list-group-flush'>
            {
                todos.map((todo, i) => (
                    
                    <TodoListItem 
                    key={ todo.id }
                    todo={ todo } 
                    index={ i } 
                    handelToggle={handelToggle} 
                    handelDelete={handelDelete}/>
                    
                ))
            }

        </ul>
    )
}

// TodoList.propTypes = {

// }

