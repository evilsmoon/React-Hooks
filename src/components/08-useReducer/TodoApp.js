import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer';
// import { useForm } from '../../hooks/useForm';
import { TodoList } from './TodoList';

import './styles.css'
import { TodoAdd } from './TodoAdd';

const init = ()=> {
    // return [{
    // id: new Date().getTime(),
    // desc: 'aprendiendo React',
    // done : false
    // }];
    return JSON.parse(localStorage.getItem('todos')) || [];
}
export const TodoApp = () => {
    
    
    const [todos,dispatch] = useReducer(todoReducer, [],init);
    
    

    useEffect(() => {
        localStorage.setItem('todos',JSON.stringify(todos));
    }, [todos]);


    const handelDelete = ( todoId ) =>{
        const action = {
            type: 'delete',
            payload:todoId 

            
        }
        dispatch(action);
   
    }

    const handelToggle = ( todoId ) =>{

        dispatch({
            type:'toggle',
            payload:todoId
        })
    }

    const handelAddTodo = ( newToDo) =>{
        dispatch({
            type: 'add',
            payload: newToDo
        } )
    }
    

    return (
        <div>
            <h1>Todo <small>{todos.length}</small></h1>
            <hr/>
        <div className='row'>
            <div className='col-7'>
            <>
            <TodoList 
                    todos={todos} 
                    handelDelete={handelDelete} 
                    handelToggle={handelToggle}/>
            </>

            </div>
            <div className='col-5'>
                <TodoAdd 
                    handelAddTodo={handelAddTodo}/>
            </div>
        </div>
       
        </div>
    )
}
