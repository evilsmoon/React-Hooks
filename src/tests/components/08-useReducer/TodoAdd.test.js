import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import React from 'react';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';
describe('test TodoAdd', () => {

    const handelAddTodo = jest.fn()

    const wrapper = shallow(<TodoAdd

        handelAddTodo={handelAddTodo}

    />)

    test('debe de mostrar correctamente <TodoAdd/>', () => {



        expect(wrapper).toMatchSnapshot();

    });

    test('no debe de llamar handeltodoadd', () => {

        const FormSubmit = wrapper.find('form').prop('onSubmit');

        // console.log(FormSubmit);
        FormSubmit({ preventDefault() { } });
        expect(handelAddTodo).toBeCalledTimes(0)

    });
    test('debe de llamar handelAddTodo', () => {

        const value = "aprender react"

        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'description'
            }

        })
        const FormSubmit = wrapper.find('form').prop('onSubmit');

        FormSubmit({ preventDefault() { } });
        expect( handelAddTodo ).toHaveBeenCalledTimes( 1 )
        expect( handelAddTodo ).toHaveBeenCalledWith( expect.any(Object) );
        expect( handelAddTodo ).toHaveBeenCalledWith( {
            id :expect.any(Number),
            desc:value,
            done:false,

        } );
        expect( wrapper.find('input').prop('value') ).toBe('')


    })




})              
