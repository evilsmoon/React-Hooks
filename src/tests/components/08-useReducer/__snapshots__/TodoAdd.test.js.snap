// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test TodoAdd debe de mostrar correctamente <TodoAdd/> 1`] = `
<Fragment>
  <h4>
    add ToDo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="add.."
      type="text"
      value=""
    />
    <button
      className="btn btn-primary mt-1 btn-block"
      type="submit"
    >
      Add
    </button>
  </form>
</Fragment>
`;
